libs/javascript/@local/tsconfig/legacy-base-tsconfig-to-refactor.json

```diff
@@ -10,17 +10,7 @@
     "resolveJsonModule": true,
     "skipLibCheck": true,
     "strict": true,
-    "target": "es2020",
-    "baseUrl": ".",
-    "paths": {
-      "@hashintel/hash-shared/*": ["../../../../packages/hash/shared/src/*"],
```

Comment:
I think this was added originally to make code intelligence work? Need to check how this behaves in WebStorm

Reply:
Yep, please do. I believe that we added `paths` because `exports` were not supported by some tooling or we did not figure out how to set it. VSCode understands `exports`, Node.js and Next.js too. That's a common standard, while `paths` is a TS-specific ‘hack’.

Reply:
Done some research – this was added in https://github.com/hashintel/dev/pull/436 

On this branch now, when I try to autoimport `generateTypeId`, for example, the import that gets added is:

```ts
import { generateTypeId } from "@hashintel/hash-shared/src/types";
```

This is not correct. See that it has the `src` in the path, when it shouldn't. 

See the video below. This error doesn't occur on main.

https://www.loom.com/share/6d7e93edf7704d56866de38b560a3e20

Reply:
That's quite odd. I installed WebStorm and could reproduce it. What’s really strange is that there are two import suggestions. The default one suggests the wrong path (`@hashintel/hash-shared/src/types`) but the one from TS suggests the right path (`@hashintel/hash-shared/types`):

<img width="1745" alt="Screenshot 2022-11-17 at 12 11 45" src="https://user-images.githubusercontent.com/608862/202444940-d23070cb-b03f-4296-a283-e0ba75e9f3ac.png">

Reply:
The same applies to WebStorm 2022.3 EAP (Build #WS-223.7571.1, built on November 8, 2022). The UI looks a bit different but the problem is the same:

<img width="820" alt="Screenshot 2022-11-17 at 12 39 56" src="https://user-images.githubusercontent.com/608862/202450559-fa185227-f9c2-4f76-8625-3ba3fcffe5a8.png">

Reply:
If I disable the built-in inspection for imports:

<img width="893" alt="Screenshot 2022-11-17 at 12 46 13" src="https://user-images.githubusercontent.com/608862/202450779-fcf7bdf6-1dae-45f8-b88a-bcc3e4977fcf.png">

import autosuggestion seems to be working correctly:

<img width="665" alt="Screenshot 2022-11-17 at 12 46 43" src="https://user-images.githubusercontent.com/608862/202450923-f1e4d61b-ffba-422e-8bdf-47216ea97db8.png">

<img width="859" alt="Screenshot 2022-11-17 at 12 46 56" src="https://user-images.githubusercontent.com/608862/202450927-049bbf24-4ccf-4e00-afc8-3a3abf58cb40.png">

My conclusion is that we are dealing with an internal WebStorm bug. Maybe there's also a way to configure the inspector, but I haven’t gone deeper as there is a workaround. Using TS server for imports makes sense because it works equally for everyone, regardless of the UI being used.

Reply:
For future travellers, some relevant webstorm issues:

- https://youtrack.jetbrains.com/issue/WEB-40509/Support-exports-field-of-package.json
- https://youtrack.jetbrains.com/issue/WEB-54095

Reply:
I see, hope that WebStorm supports `exports` properly soon!

Which other consequences would `Unresolved TypeScript function` and `Unresolved TypeScript variable` bring? It feels that disabling both of them solves the problem:

<img width="844" alt="Screenshot 2022-11-17 at 15 00 53" src="https://user-images.githubusercontent.com/608862/202481140-7b520cea-0ac8-432b-968a-f3994ec250e0.png">

I don’t mind keeping `paths` in our base config for now, but I’m not of big fan of this approach TBH. It leaves us with two ‘sources of truth’ which can eventually divert from each other over create odd DX bugs. When you say that WebStorm’s auto imports are more powerful, which specific use cases do you mean? I’d like to understand practical pros of bringing `paths` back.

Reply:
Ultimately I shouldn't have to be turning off standard editor features to make our codebase work. Exports is not a widely enough support feature yet to be relying on it on its own, so we need to keep paths. I'm aware there's an issue with "two sources of truth" but that is a consequence of being in a transitionary period where different tools support different features. 

These inspections are useful. Without these inspections you have to wait for the TS server to report these problems, which is slower. You also lose some suggestions

This is the menu with the inspections

<img width="423" alt="image" src="https://user-images.githubusercontent.com/719814/202500059-7e9bd23b-d56e-4bcf-967b-a67ff86c215a.png">

This is the menu without

<img width="452" alt="image" src="https://user-images.githubusercontent.com/719814/202500152-4337e658-3caa-4a1f-96fe-d8d9553dc8a4.png">

The TS "add missing function declaration" also isn't working. 

When you have multiple missing imports, this is the menu with inspections

<img width="736" alt="image" src="https://user-images.githubusercontent.com/719814/202500666-0e05f69b-a538-4b30-8cfc-0f479b01797c.png">

Without the inspections, this is the menu

<img width="740" alt="image" src="https://user-images.githubusercontent.com/719814/202500796-a38b71a1-bdf4-41d7-b121-90d012db6c6e.png">

You lose the one click "add all ambiguous imports". Not to mention that importing with TS doesn't respect import order whereas webstorm does.

For me, disabling webstorm features to avoid needing to use paths is not a good trade off.

Reply:
Fair point. I brought back paths, but still got `@hashintel/hash-shared/src/types` in WebStorm. Could you check it on your side?

Maybe we should just not bother with path mapping. We already have `@hashintel/hash-api/src/model/util`, maybe we can just go for `@hashintel/hash-shared/src/types` too till WebStorm autocompletion is fixed. We will need path mapping in `exports` later when code is in `src/*` and the built version is in `dist/*`. That's a necessity if some of our workspaces end up published on `npm`.

Reply:
It is working as expected on your branch now. I don't think we need to add "src" to paths at this point in time, that would be a very disruptive change.

Reply:
Nice! OK, let's get this in as is and if we see any further issues with import paths, we can just add `/src/` everywhere as a temporary measure. This will also allow us to remove `tsconfig.json` → `paths`, thus removing the redundant ‘source of truth’.